### ISSUE TITLE
[Feature] Add run_id, phase and timestamp to Socket.IO training events

### ISSUE THREAD

**Issue by Shivampal157** [c2siorg/tensormap#274]
**Difficulty:** Medium
**Area:** full stack

## Description
The message-index half of this request has landed and is not in scope: `resultValues[parseInt(resp.test)]` is gone from `containers/Training/Training.jsx`, the live path now runs through `MetricsCallback`, which persists per-epoch metrics and emits typed `{type: "metrics" | "status"}` payloads into a Socket.IO room keyed by `job_id`, and `subscribeToJob()` filters on `data.type`. What survives is the payload contract itself. A metrics event at HEAD carries only `{type, epoch, loss, accuracy, val_loss, val_accuracy}`: no `run_id` or `job_id`, no `phase`, no `batch`, no step count, no timestamp, no nested `metrics` object, and no contract version. Room isolation covers the single-subscription case, but a client holding two subscriptions on the one shared socket cannot attribute an event to its run, and `useTrainingMetrics` already reads a `data.started_at` on catch-up that the server never sends. If a maintainer judges room isolation sufficient and the extra fields unnecessary, the right outcome is to say so here and close this, not to leave it open.

## Acceptance Criteria
- [ ] Every payload emitted on the training listener carries the `job_id` it belongs to, so a consumer can attribute an event without relying on which room delivered it.
- [ ] Metrics events carry a `phase` distinguishing train from validation, and an ISO-8601 `timestamp`.
- [ ] Metrics events carry the within-epoch position (`batch` and `steps`) so progress can advance inside an epoch, not only between epochs.
- [ ] Every payload carries a contract version, and `subscribeToJob()` drops payloads whose version it does not understand.
- [ ] `_load_catchup()` includes `started_at`, which `useTrainingMetrics` already tries to read and currently never receives.
- [ ] The terminal `status` payloads from `on_train_end()` and from the `model_run()` failure path use the same envelope as the metrics events.
- [ ] `types/training.ts` and the `socketService` / `useTrainingMetrics` tests are updated to cover the new envelope.

## Resources
- `tensormap-backend/app/callbacks/metrics_callback.py:74-104` - `on_epoch_end()` builds and emits the per-epoch payload
- `tensormap-backend/app/callbacks/metrics_callback.py:137-138` - the terminal `status` emit
- `tensormap-backend/app/callbacks/metrics_callback.py:25-41` - `schedule_room_emit()`, the single room-emit path shared by the callback and the failure handler
- `tensormap-backend/app/services/model_run.py:195` - the failure-path `status` emit, which must use the same envelope
- `tensormap-backend/app/socketio_instance.py:31-63` - `_load_catchup()` and `subscribe_job()`; the catch-up payload has no `started_at`
- `tensormap-frontend/src/services/socketService.js:42-67` - `subscribeToJob()`, which filters purely on `data.type`
- `tensormap-frontend/src/hooks/useTrainingMetrics.ts:70-90` - the consumer; line 76 reads `data.started_at`, which is never emitted
- `tensormap-frontend/src/types/training.ts:6-12` - `EpochMetric`, the typed shape to extend

<details>
<summary>Original report by @Shivampal157 (2026-03-31)</summary>

## Problem

Current training progress events are emitted as loosely structured payloads (`message` + numeric `test` index), and the frontend stores them by array index.

This makes the contract fragile and hard to extend:
- UI behavior depends on parsing free-form strings like `"Starting"` / `"Finish"`
- No stable `run_id` in event payload to scope updates
- No explicit epoch/batch/metric fields for charting
- Hard to support multi-run comparison and robust progress visualizations

## Why this matters

A stable telemetry contract is required for:
- real-time training charts
- run comparison dashboard
- reliable retry/reconnect handling
- future tuning/interpretability workflows

String-based progress messages are hard to validate and can silently break UI.

## Current behavior (observed)

- Backend emits generic message events from `CustomProgressBar`
- Frontend in `Training.jsx` maps updates using `resultValues[parseInt(resp.test)]`
- Progress state is not strongly typed and not run-scoped

## Proposed solution

Introduce versioned structured events for training:

- `training_started`
- `training_update`
- `training_complete`
- `training_error`

Each event should include a typed payload, e.g.:

```json
{
  "run_id": "<uuid-or-id>",
  "phase": "train|eval",
  "epoch": 3,
  "batch": 12,
  "steps": 100,
  "metrics": {
    "loss": 0.42,
    "accuracy": 0.88,
    "val_loss": 0.50,
    "val_accuracy": 0.84
  },
  "message": "optional human-readable message",
  "timestamp": "ISO-8601"
}

</details>
